### DISCUSSION TITLE
Code discussion

### DISCUSSION THREAD

**Discussion by ssrlive** [narrowlink/ipstack#43] [General]
The `contains` function

https://github.com/narrowlink/ipstack/blob/31ca1b3d62d8372061615faf4fd1656b2fc4428e/src/stream/tcb.rs#L221-L223

can change to this?
```rust
    pub(crate) fn contains(&self, seq: u32) -> bool {
        self.seq <= seq && seq < self.seq + self.payload.len() as u32
    }
```

**Comment by ssrlive** [comment]
About `check_pkt_type` function

https://github.com/narrowlink/ipstack/blob/31ca1b3d62d8372061615faf4fd1656b2fc4428e/src/stream/tcb.rs#L144-L154

Is there a potential issue is that this function might not handle certain edge cases correctly?

Specifically, when `incoming_packet.inner().acknowledgment_number` equals `self.seq` and `self.seq.saturating_sub(incoming_packet.inner().acknowledgment_number)` equals 0, this function returns `PacketStatus::Invalid`. However, this situation might actually represent a valid packet, just that its acknowledgment number happens to equal the current sequence number.

If we change this condition to `incoming_packet.inner().acknowledgment_number != self.seq || self.seq.saturating_sub(incoming_packet.inner().acknowledgment_number) > 0`, so that it only returns `PacketStatus::Invalid` when the acknowledgment number does not equal the current sequence number and the distance between the current sequence number and the acknowledgment number is greater than 0.

**Reply by SajjadPourali** [reply]
This function is one of the most critical functions, and we have to be very careful with it. I used saturating_sub because TCP sequence/ack numbers can roll over. So, 0 here doesn't mean they are equal always; it means the acknowledgment received is acknowledging something beyond what we sent, which doesn't exist. I don't see the situation where it throws an invalid error when a valid packet arrives, but the condition that you suggested simply invalidates the actual valid packets. For example, if the current sequence is 20 (the pointer to the last byte that was sent) and the peer acknowledges 10, then 20 - 10 = 10, which is larger than 0, and what you suggested considers it invalid, while it is a valid acknowledge number.

**Comment by ssrlive** [comment]
Here some ACK packet will discarded.

https://github.com/narrowlink/ipstack/blob/a8afef853590a1a267c304592c6f424f53970f3e/src/stream/tcp.rs#L86-L92

We can watch it with this
```rust
        if tcp.inner().syn {
            return Ok(stream);
        }
        if !tcp.inner().rst {
            _ = pkt_sender.send(stream.create_rev_packet(RST | ACK, TTL, None, Vec::new())?); 
        }
        if tcp.inner().ack {
            log::info!("Invalid TCP packet with ACK flag");
        }
        Err(IpStackError::InvalidTcpPacket)
```

**Reply by SajjadPourali** [reply]
There are also other situations where we may receive invalid packets, e.g., null flags or christmas tree packets. Let's omit logging invalid packets at this time. We may also print them in the debug mode.

**Comment by SajjadPourali** [comment]
> The `contains` function
> 
> https://github.com/narrowlink/ipstack/blob/31ca1b3d62d8372061615faf4fd1656b2fc4428e/src/stream/tcb.rs#L221-L223
> 
> can change to this?
> 
> ```rust
>     pub(crate) fn contains(&self, seq: u32) -> bool {
>         self.seq <= seq && seq < self.seq + self.payload.len() as u32
>     }
> ```

`self.seq` is the sequence pointer at the time the packet is added. So, the offset of the first packet in the payload is `seq + 1`. In other words, we should use `self.seq < seq` rather than `self.seq <= seq`. The same applies to the second condition.

**Comment by ssrlive** [comment]
Here is my reading code results, Please review it. https://github.com/narrowlink/ipstack/pull/41

**Comment by ssrlive** [comment]
Some pictures.

![tcp-connect](https://github.com/user-attachments/assets/ad5c7a37-aab8-4ee8-a376-966aee571363)

![tcp-established](https://github.com/user-attachments/assets/61f5f427-6b68-4d84-b81b-a04b544044bb)

![tcp-exiting](https://github.com/user-attachments/assets/d667c69e-a4de-4dd2-ad6c-6f1d91d44327)

**Reply by SajjadPourali** [reply]
Could you provide a brief explanation?

**Comment by ssrlive** [comment]
It's just for my own reference to reconstruct the code. These pictures are
three handshakes and four farewell waves of the TCP protocol.

**Comment by ssrlive** [comment]
Here is a diagram of the refactored code.

![image](https://github.com/user-attachments/assets/2b8b6049-e238-4040-94f0-947e37f56bbd)
